tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11236,1074 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+    return;
+  }
+
+  DeterministicNumberGenerator<T> NumberGenerator(1337);
+  std::vector<T> VecInput1 = {};
+  std::vector<T> VecInput2 = {};
+  const T ScalarInput = NumberGenerator.generate();
+  const bool IsVectorBinaryOp = TestConfig.IsBinaryOp && !TestConfig.IsScalarOp;
+
+  // Fill the vector inputs with values.
+  for (size_t i = 0; i < N; i++) {
+    // Always generate input. Duh.
+    VecInput1.push_back(NumberGenerator.generate());
+
+    if (IsVectorBinaryOp) {
+      VecInput2.push_back(NumberGenerator.generate());
+    }
+  }
+
+  // We pass these values into the shader and they're requried to compile. So
+  // they need to set to something.
+  T ClampArgC = 0;
+  T ClampArgT = 0;
+  if (TestConfig.IntrinsicString == "testClamp") {
+    ClampArgC = NumberGenerator.generate();
+    ClampArgT = NumberGenerator.generate();
+    while (ClampArgC >= ClampArgT) {
+      // Generate a new value for ClampArgC. It needs to be smaller than
+      // ClampArgT.
+      ClampArgT = NumberGenerator.generate();
+    }
+  }
+
+  std::vector<T> VecExpected = {};
+  for (size_t Index = 0; Index < VecInput1.size(); Index++) {
+    if (TestConfig.IsBinaryOp) {
+      T Input1 = VecInput1[Index];
+      T Input2 = TestConfig.IsScalarOp ? ScalarInput : VecInput2[Index];
+      if (TestConfig.OperatorString == "*")
+        VecExpected.push_back(Input1 * Input2);
+      else if (TestConfig.OperatorString == "+")
+        VecExpected.push_back(Input1 + Input2);
+      else if (TestConfig.OperatorString == ",") {
+        VERIFY_IS_TRUE(TestConfig.IntrinsicString != "",
+                       "Expecting intrinsic string");
+        if (TestConfig.IntrinsicString == "min")
+          VecExpected.push_back(std::min<T>(Input1, Input2));
+        else if (TestConfig.IntrinsicString == "max")
+          VecExpected.push_back(std::max<T>(Input1, Input2));
+        else
+          LogErrorFmt(L"Unrecognized BinaryOp intrinsic string: %s",
+                      TestConfig.IntrinsicString.c_str());
+      } else
+        LogErrorFmt(
+            L"Don't know how to compute expected value for operatorString: %s",
+            TestConfig.OperatorString.c_str());
+    } else // Unary op logic
+    {
+      if (TestConfig.IntrinsicString == "testClamp")
+        VecExpected.push_back(
+            std::clamp(VecInput1[Index], ClampArgC, ClampArgT));
+      else if (TestConfig.IntrinsicString == "testInitialize")
+        VecExpected.push_back(VecInput1[Index]);
+      else
+        LogErrorFmt(L"Unrecognized intrinsic string: %s",
+                    TestConfig.IntrinsicString.c_str());
+    }
+  }
+
+  // 1,2, and 3 are required to compile the shader.
+  // 1. -DOPERATOR : "*" "+" "," "-" "/" etc.
+  // 2. -DOPERAND2: : l.VecInput2 or scalarInput, depending on the test.
+  // 3. -DCLAMP_ARGC, -DCLAMP_ARG2 : clamp arg 1 and 2. Required to compile.
```

Comment:
Just a curious thought, up to you if you want to do this.
You might try ifdefing the testClamp function in the shader if the two defs don't exist, and then you can conditionally add the clamp args in 12245. This way there are less options to the compiler that aren't used. This works too. Then 3. isn't required to compile.

Reply:
Adjusted in latest revision.